numpy/lib/function_base.py

```diff
@@ -1912,10 +1911,16 @@ def diff(a, n=1, axis=-1):
     slice2[axis] = slice(None, -1)
     slice1 = tuple(slice1)
     slice2 = tuple(slice2)
+
+    if a.dtype == np.bool_:
+        da = a[slice1] ^ a[slice2]
```

Comment:
> Could you explain a bit more.

Test would become `a.dtype == np.bool_ or a.dtype.fields`, and would enable `np.diff(struct_array)`, meaning `adjacent entries are different.

Of course, it would be weird to not also support this for float arrays, suggesting that the "adjacent entries are different" operation should be given a different name to `diff`, and `diff(bool_array)` should DeprecationWarning, recommending the new name